contracts/src/Payment.sol

```diff
@@ -159,4 +170,54 @@ contract Payment {
         Channel storage channel = channels[publisher][provider][podId];
         return channel.withdrawnByProvider;
     }
+
+    // authorize other addresses to create subChannels
+    function authorize(address _authorized, address provider, bytes32 podId) public {
+        Channel storage channel = channels[msg.sender][provider][podId];
+        if (channel.investedByPublisher == 0) revert DoesNotExist();
+        channel.authorized[_authorized] = true;
+    }
+
+    function isAuthorized(address publisher, address provider, bytes32 podId, address _address)
+        public
+        view
+        returns (bool)
+    {
+        return channels[publisher][provider][podId].authorized[_address];
+    }
+
+    // create a subChannel from a main channel
+    function createSubChannel(
+        address publisher,
+        address provider,
+        bytes32 podId,
+        address newProvider,
+        bytes32 newPodId,
+        uint256 amount
+    ) public {
+        Channel storage channel = channels[publisher][provider][podId];
+        // Ensure the channel exists
+        if (channel.investedByPublisher == 0) revert DoesNotExist();
+
+        // Check if the caller is authorized
+        if (!channel.authorized[msg.sender]) revert NotAuthorized();
+
+        // Ensure there is enough invested by the publisher
+        if (channel.investedByPublisher < amount) revert InsufficientFunds();
+
+        // Deduct the amount from the main channel's funds
+        channel.investedByPublisher -= amount;
```

Comment:
Decreasing the amount invested by the publisher without requiring the channel to be unlocked first allows the publisher to drain the channel of any funds within it, by creating a subchannel with all the funds they can, then unlocking that subchannel, and withdrawing from it at a leisurely pace, without the provider able to submit a `withdraw()` call.

Also, since the code never checks whether `channel.investedByPublisher - channel.withdrawnByProvider < amount` (see, e.g. line 121, `leftoverFunds = ..`), it's possible to withdraw more than was initially invested into a channel, thus draining the whole contract of funds.

Reply:
Fixed as well